File: sdk/sdk-js/src/methods/archive.ts
import {
    ArchiveBalanceCheckpoint,
    ArchiveBalanceInfo,
    ArchiveHealth,
    ArchiveExportRequest,
    ArchiveExportResponse,
    ArchiveManifest,
    ArchiveMigration,
    ArchiveMigrationBatch,
    ArchiveMigrationBatchRequest,
    ArchiveMigrationRequest,
    ArchiveObject,
    ArchiveObjectMetadata,
    ArchiveQuery,
    ArchiveRangeInfo,
    ArchiveStoragePoolHealth,
    ArchiveStoragePool,
    ArchiveVerificationResult,
    Entry,
    EnumerationResult,
    RequestHistoryEntry,
    RequestHistoryQuery,
    RequestHistorySummary
} from '../models';
import { HttpClient } from '../http-client';
import { URLSearchParams } from 'url';

function appendQuery(path: string, query?: ArchiveQuery): string {
    if (!query) return path;

    const params = new URLSearchParams();
    Object.entries(query).forEach(([key, value]) => {
        if (value !== null && value !== undefined && value !== '') {
            if (Array.isArray(value)) {
                params.append(key, value.join(','));
            } else if (typeof value === 'object') {
                params.append(key, Object.entries(value).map(([itemKey, itemValue]) => `${itemKey}=${itemValue}`).join(','));
            } else {
                params.append(key, String(value));
            }
        }
    });

    const queryString = params.toString();
    return queryString ? `${path}?${queryString}` : path;
}

function appendRequestHistoryQuery(path: string, query?: RequestHistoryQuery): string {
    if (!query) return path;

    const params = new URLSearchParams();
    Object.entries(query).forEach(([key, value]) => {
        if (value !== null && value !== undefined && value !== '') {
            params.append(key.charAt(0).toLowerCase() + key.slice(1), String(value));
        }
    });

    const queryString = params.toString();
    return queryString ? `${path}?${queryString}` : path;
}

/**
 * Archive Server operations for cold data and archive metadata.
 */
export class ArchiveMethods {
    private readonly client: HttpClient;

    constructor(client: HttpClient) {
        this.client = client;
    }

    async health(): Promise<ArchiveHealth> {
        const response = await this.client.get<ArchiveHealth>('/v1/health');
        return response.Data || { Healthy: false };
    }

    async ranges(query?: ArchiveQuery): Promise<ArchiveRangeInfo[]> {
        const response = await this.client.get<ArchiveRangeInfo[]>(appendQuery('/v1/archive/ranges', query));
        return response.Data || [];
    }

    async manifests(query?: ArchiveQuery): Promise<ArchiveManifest[]> {
        const response = await this.client.get<ArchiveManifest[]>(appendQuery('/v1/archive/manifests', query));
        return response.Data || [];
    }

    async manifest(manifestId: string): Promise<ArchiveManifest> {
        const response = await this.client.get<ArchiveManifest>(`/v1/archive/manifests/${encodeURIComponent(manifestId)}`);
        if (!response.Data) throw new Error('No manifest returned from Archive Server');
        return response.Data;
    }

    async manifestObjects(manifestId: string, query?: ArchiveQuery): Promise<ArchiveObject[]> {
        const response = await this.client.get<ArchiveObject[]>(
            appendQuery(`/v1/archive/manifests/${encodeURIComponent(manifestId)}/objects`, query)
        );
        return response.Data || [];
    }

    async manifestCheckpoints(manifestId: string, query?: ArchiveQuery): Promise<ArchiveBalanceCheckpoint[]> {
        const response = await this.client.get<ArchiveBalanceCheckpoint[]>(
            appendQuery(`/v1/archive/manifests/${encodeURIComponent(manifestId)}/checkpoints`, query)
        );
        return response.Data || [];
    }

    async verifyManifest(manifestId: string): Promise<void> {
        await this.client.post(`/v1/archive/manifests/${encodeURIComponent(manifestId)}/verify`);
    }

    async quarantineManifest(manifestId: string): Promise<void> {
        await this.client.post(`/v1/archive/manifests/${encodeURIComponent(manifestId)}/quarantine`);
    }

    async supersedeManifest(manifestId: string): Promise<void> {
        await this.client.post(`/v1/archive/manifests/${encodeURIComponent(manifestId)}/supersede`);
    }

    async storagePools(query?: ArchiveQuery): Promise<ArchiveStoragePool[]> {
        const response = await this.client.get<ArchiveStoragePool[]>(appendQuery('/v1/archive/storage-pools', query));
        return response.Data || [];
    }

    async storagePoolHealth(storagePoolId: string): Promise<ArchiveStoragePoolHealth> {
        const response = await this.client.get<ArchiveStoragePoolHealth>(`/v1/archive/storage-pools/${encodeURIComponent(storagePoolId)}/health`);
        return response.Data || { Healthy: false, StoragePoolId: storagePoolId };
    }

    async migrations(query?: ArchiveQuery): Promise<ArchiveMigration[]> {
        const response = await this.client.get<ArchiveMigration[]>(appendQuery('/v1/archive/migrations', query));
        return response.Data || [];
    }

    async migration(migrationId: string): Promise<ArchiveMigration> {
        const response = await this.client.get<ArchiveMigration>(`/v1/archive/migrations/${encodeURIComponent(migrationId)}`);
        if (!response.Data) throw new Error('No migration returned from Archive Server');
        return response.Data;
    }

    async migrationBatches(migrationId: string, query?: ArchiveQuery): Promise<ArchiveMigrationBatch[]> {
        const response = await this.client.get<ArchiveMigrationBatch[]>(
            appendQuery(`/v1/archive/migrations/${encodeURIComponent(migrationId)}/batches`, query)
        );
        return response.Data || [];
    }

    async createMigration(request: ArchiveMigrationRequest): Promise<ArchiveMigration> {
        const response = await this.client.post<ArchiveMigration>('/v1/archive/migrations', request);
        if (!response.Data) throw new Error('No migration returned from Archive Server');
        return response.Data;
    }

    async createMigrationBatch(migrationId: string, request: ArchiveMigrationBatchRequest): Promise<ArchiveMigrationBatch> {
        const response = await this.client.post<ArchiveMigrationBatch>(
            `/v1/archive/migrations/${encodeURIComponent(migrationId)}/batches`,
            request
        );
        if (!response.Data) throw new Error('No migration batch returned from Archive Server');
        return response.Data;
    }

    async uploadMigrationBatchContent(
        migrationId: string,
        batchId: string,
        content: Buffer | Uint8Array | string,
        contentHashSha256?: string,
        contentType: string = 'application/gzip'
    ): Promise<ArchiveMigrationBatch> {
        const response = await this.client.putRaw<ArchiveMigrationBatch>(
            `/v1/archive/migrations/${encodeURIComponent(migrationId)}/batches/${encodeURIComponent(batchId)}/content`,
            content,
            contentType,
            contentHashSha256 ? { 'x-content-sha256': contentHashSha256 } : undefined
        );
        if (!response.Data) throw new Error('No migration batch returned from Archive Server');
        return response.Data;
    }

    async sealMigration(migrationId: string): Promise<ArchiveMigration> {
        const response = await this.client.post<ArchiveMigration>(`/v1/archive/migrations/${encodeURIComponent(migrationId)}/seal`);
        if (!response.Data) throw new Error('No migration returned from Archive Server');
        return response.Data;
    }

    async commitMigration(migrationId: string): Promise<ArchiveManifest> {
        const response = await this.client.post<ArchiveManifest>(`/v1/archive/migrations/${encodeURIComponent(migrationId)}/commit`);
        if (!response.Data) throw new Error('No manifest returned from Archive Server');
        return response.Data;
    }

    async abortMigration(migrationId: string): Promise<ArchiveMigration> {
        const response = await this.client.post<ArchiveMigration>(`/v1/archive/migrations/${encodeURIComponent(migrationId)}/abort`);
        if (!response.Data) throw new Error('No migration returned from Archive Server');
        return response.Data;
    }

    async exportEntries(request: ArchiveExportRequest): Promise<ArchiveExportResponse> {
        const response = await this.client.post<ArchiveExportResponse>('/v1/archive/exports/entries', request);
        return response.Data || { RowsExported: 0, BytesUploaded: 0, ActiveCleanupExecuted: false, ActiveCleanupRowsDeleted: 0, Batches: [] };
    }

    async exportRequestHistory(request: ArchiveExportRequest): Promise<ArchiveExportResponse> {
        const response = await this.client.post<ArchiveExportResponse>('/v1/archive/exports/request-history', request);
        return response.Data || { RowsExported: 0, BytesUploaded: 0, ActiveCleanupExecuted: false, ActiveCleanupRowsDeleted: 0, Batches: [] };
    }

    async exportTenantAccountEntries(tenantId: string, accountId: string, request?: ArchiveExportRequest): Promise<ArchiveExportResponse> {
        const response = await this.client.post<ArchiveExportResponse>(
            `/v1/tenants/${encodeURIComponent(tenantId)}/accounts/${encodeURIComponent(accountId)}/archive/export`,
            request || {}
        );
        return response.Data || { RowsExported: 0, BytesUploaded: 0, ActiveCleanupExecuted: false, ActiveCleanupRowsDeleted: 0, Batches: [] };
    }

    async entries(accountId: string, query?: ArchiveQuery): Promise<EnumerationResult<Entry>> {
        const response = await this.client.get<EnumerationResult<Entry>>(
            appendQuery(`/v1/archive/accounts/${encodeURIComponent(accountId)}/entries`, query)
        );
        return response.Data || { TotalRecords: 0, RecordsRemaining: 0, EndOfResults: true, Objects: [] };
    }

    async tenantEntries(tenantId: string, accountId: string, query?: ArchiveQuery): Promise<EnumerationResult<Entry>> {
        const response = await this.client.get<EnumerationResult<Entry>>(
            appendQuery(`/v1/tenants/${encodeURIComponent(tenantId)}/accounts/${encodeURIComponent(accountId)}/entries`, query)
        );
        return response.Data || { TotalRecords: 0, RecordsRemaining: 0, EndOfResults: true, Objects: [] };
    }

    async balanceAsOf(accountId: string, asOfUtc: string | Date): Promise<ArchiveBalanceInfo> {
        const asOf = asOfUtc instanceof Date ? asOfUtc.toISOString() : asOfUtc;
        const response = await this.client.get<ArchiveBalanceInfo>(
            `/v1/archive/accounts/${encodeURIComponent(accountId)}/balance/asof?asOf=${encodeURIComponent(asOf)}`
        );
        if (!response.Data) throw new Error('No archived balance returned from Archive Server');
        return response.Data;
    }

    async tenantBalanceAsOf(tenantId: string, accountId: string, asOfUtc: string | Date): Promise<ArchiveBalanceInfo> {
        const asOf = asOfUtc instanceof Date ? asOfUtc.toISOString() : asOfUtc;
        const response = await this.client.get<ArchiveBalanceInfo>(
            `/v1/tenants/${encodeURIComponent(tenantId)}/accounts/${encodeURIComponent(accountId)}/balance/asof?asOf=${encodeURIComponent(asOf)}`
        );
        if (!response.Data) throw new Error('No archived balance returned from Archive Server');
        return response.Data;
    }

    async verifyAccount(accountId: string, query?: ArchiveQuery): Promise<ArchiveVerificationResult> {
        const response = await this.client.get<ArchiveVerificationResult>(
            appendQuery(`/v1/archive/accounts/${encodeURIComponent(accountId)}/verify`, query)
        );
        return response.Data || { TenantId: '', AccountId: accountId, IsValid: true, CheckedManifests: 0, CheckedObjects: 0, CheckedBalanceCheckpoints: 0, Details: [], Errors: [] };
    }

    async verifyTenantAccount(tenantId: string, accountId: string, query?: ArchiveQuery): Promise<ArchiveVerificationResult> {
        const response = await this.client.get<ArchiveVerificationResult>(
            appendQuery(`/v1/tenants/${encodeURIComponent(tenantId)}/accounts/${encodeURIComponent(accountId)}/verify`, query)
        );
        return response.Data || { TenantId: tenantId, AccountId: accountId, IsValid: true, CheckedManifests: 0, CheckedObjects: 0, CheckedBalanceCheckpoints: 0, Details: [], Errors: [] };
    }

    async objectMetadata(objectId: string): Promise<ArchiveObjectMetadata> {
        const response = await this.client.get<ArchiveObjectMetadata>(`/v1/archive/objects/${encodeURIComponent(objectId)}/metadata`);
        if (!response.Data) throw new Error('No archive object metadata returned from Archive Server');
        return response.Data;
    }

    async requestHistory(query?: RequestHistoryQuery): Promise<EnumerationResult<RequestHistoryEntry>> {
        const response = await this.client.get<EnumerationResult<RequestHistoryEntry>>(appendRequestHistoryQuery('/v1/request-history', query));
        return response.Data || { TotalRecords: 0, RecordsRemaining: 0, EndOfResults: true, Objects: [] };
    }

    async requestHistorySummary(query?: RequestHistoryQuery): Promise<RequestHistorySummary> {
        const response = await this.client.get<RequestHistorySummary>(appendRequestHistoryQuery('/v1/request-history/summary', query));
        return response.Data || { TotalCount: 0, TotalSuccess: 0, TotalFailure: 0, AverageDurationMs: 0, Buckets: [] };
    }

    async requestHistoryEntry(id: string, query?: RequestHistoryQuery): Promise<RequestHistoryEntry> {
        const response = await this.client.get<RequestHistoryEntry>(appendRequestHistoryQuery(`/v1/request-history/${encodeURIComponent(id)}`, query));
        if (!response.Data) throw new Error('No request history entry returned from Archive Server');
        return response.Data;
    }

    async archiveServerRequestHistory(query?: RequestHistoryQuery): Promise<EnumerationResult<RequestHistoryEntry>> {
        const response = await this.client.get<EnumerationResult<RequestHistoryEntry>>(appendRequestHistoryQuery('/v1/archive-server/request-history', query));
        return response.Data || { TotalRecords: 0, RecordsRemaining: 0, EndOfResults: true, Objects: [] };
    }

    async archiveServerRequestHistorySummary(query?: RequestHistoryQuery): Promise<RequestHistorySummary> {
        const response = await this.client.get<RequestHistorySummary>(appendRequestHistoryQuery('/v1/archive-server/request-history/summary', query));
        return response.Data || { TotalCount: 0, TotalSuccess: 0, TotalFailure: 0, AverageDurationMs: 0, Buckets: [] };
    }

    async archiveServerRequestHistoryEntry(id: string, query?: RequestHistoryQuery): Promise<RequestHistoryEntry> {
        const response = await this.client.get<RequestHistoryEntry>(appendRequestHistoryQuery(`/v1/archive-server/request-history/${encodeURIComponent(id)}`, query));
        if (!response.Data) throw new Error('No Archive Server request history entry returned');
        return response.Data;
    }
}

File: sdk/sdk-python/netledger_sdk/methods/archive.py
"""Archive Server operations for the NetLedger SDK."""

from typing import Any, Dict, List, Optional
from urllib.parse import quote, urlencode

from ..http_client import HttpClient
from ..models import Entry, EnumerationResult


def _append_query(path: str, query: Optional[Dict[str, Any]]) -> str:
    if not query:
        return path

    parameters = []
    for key, value in query.items():
        if value is None or value == "":
            continue

        if isinstance(value, (list, tuple)):
            parameters.append((key, ",".join(str(item) for item in value)))
        elif isinstance(value, dict):
            parameters.append((key, ",".join(f"{item_key}={item_value}" for item_key, item_value in value.items())))
        else:
            parameters.append((key, str(value)))

    if not parameters:
        return path

    return f"{path}?{urlencode(parameters)}"


class ArchiveMethods:
    """Archive Server operations for cold data and archive metadata."""

    def __init__(self, client: HttpClient):
        """Initialize archive methods."""
        self._client = client

    def health(self) -> Dict[str, Any]:
        """Read Archive Server health."""
        response = self._client.get("/v1/health")
        return response.data or {"Healthy": False}

    def ranges(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List archive coverage ranges."""
        response = self._client.get(_append_query("/v1/archive/ranges", query))
        return response.data or []

    def manifests(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List archive manifests."""
        response = self._client.get(_append_query("/v1/archive/manifests", query))
        return response.data or []

    def manifest(self, manifest_id: str) -> Dict[str, Any]:
        """Read an archive manifest."""
        response = self._client.get(f"/v1/archive/manifests/{quote(manifest_id, safe='')}")
        return response.data or {}

    def manifest_objects(self, manifest_id: str, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List objects for an archive manifest."""
        path = f"/v1/archive/manifests/{quote(manifest_id, safe='')}/objects"
        response = self._client.get(_append_query(path, query))
        return response.data or []

    def manifest_checkpoints(self, manifest_id: str, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List balance checkpoints for an archive manifest."""
        path = f"/v1/archive/manifests/{quote(manifest_id, safe='')}/checkpoints"
        response = self._client.get(_append_query(path, query))
        return response.data or []

    def verify_manifest(self, manifest_id: str) -> None:
        """Verify an archive manifest."""
        self._client.post(f"/v1/archive/manifests/{quote(manifest_id, safe='')}/verify")

    def quarantine_manifest(self, manifest_id: str) -> None:
        """Quarantine an archive manifest."""
        self._client.post(f"/v1/archive/manifests/{quote(manifest_id, safe='')}/quarantine")

    def supersede_manifest(self, manifest_id: str) -> None:
        """Supersede an archive manifest."""
        self._client.post(f"/v1/archive/manifests/{quote(manifest_id, safe='')}/supersede")

    def storage_pools(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List archive storage pools."""
        response = self._client.get(_append_query("/v1/archive/storage-pools", query))
        return response.data or []

    def storage_pool_health(self, storage_pool_id: str) -> Dict[str, Any]:
        """Read archive storage pool health."""
        response = self._client.get(f"/v1/archive/storage-pools/{quote(storage_pool_id, safe='')}/health")
        return response.data or {}

    def migrations(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List archive migrations."""
        response = self._client.get(_append_query("/v1/archive/migrations", query))
        return response.data or []

    def migration(self, migration_id: str) -> Dict[str, Any]:
        """Read an archive migration."""
        response = self._client.get(f"/v1/archive/migrations/{quote(migration_id, safe='')}")
        return response.data or {}

    def migration_batches(self, migration_id: str, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List archive migration batches."""
        path = f"/v1/archive/migrations/{quote(migration_id, safe='')}/batches"
        response = self._client.get(_append_query(path, query))
        return response.data or []

    def create_migration(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Create an Archive Server migration."""
        response = self._client.post("/v1/archive/migrations", request)
        return response.data or {}

    def create_migration_batch(self, migration_id: str, request: Dict[str, Any]) -> Dict[str, Any]:
        """Create metadata for one Archive Server migration batch."""
        path = f"/v1/archive/migrations/{quote(migration_id, safe='')}/batches"
        response = self._client.post(path, request)
        return response.data or {}

    def upload_migration_batch_content(
        self,
        migration_id: str,
        batch_id: str,
        content: Any,
        content_hash_sha256: Optional[str] = None,
        content_type: str = "application/gzip"
    ) -> Dict[str, Any]:
        """Upload compressed JSONL content for one Archive Server migration batch."""
        headers = {"x-content-sha256": content_hash_sha256} if content_hash_sha256 else None
        path = f"/v1/archive/migrations/{quote(migration_id, safe='')}/batches/{quote(batch_id, safe='')}/content"
        response = self._client.put_raw(path, content, content_type, headers)
        return response.data or {}

    def seal_migration(self, migration_id: str) -> Dict[str, Any]:
        """Seal an Archive Server migration after all batches are uploaded."""
        response = self._client.post(f"/v1/archive/migrations/{quote(migration_id, safe='')}/seal")
        return response.data or {}

    def commit_migration(self, migration_id: str) -> Dict[str, Any]:
        """Commit an Archive Server migration and create its manifest."""
        response = self._client.post(f"/v1/archive/migrations/{quote(migration_id, safe='')}/commit")
        return response.data or {}

    def abort_migration(self, migration_id: str) -> Dict[str, Any]:
        """Abort an Archive Server migration and delete temporary content."""
        response = self._client.post(f"/v1/archive/migrations/{quote(migration_id, safe='')}/abort")
        return response.data or {}

    def export_entries(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Start an active NetLedger Server export of committed entries to NetLedger Archive Server."""
        response = self._client.post("/v1/archive/exports/entries", request)
        return response.data or {}

    def export_request_history(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Start an active NetLedger Server export of request history to NetLedger Archive Server."""
        response = self._client.post("/v1/archive/exports/request-history", request)
        return response.data or {}

    def export_tenant_account_entries(
        self,
        tenant_id: str,
        account_id: str,
        request: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Start an active NetLedger Server export of committed entries for one tenant account."""
        path = f"/v1/tenants/{quote(tenant_id, safe='')}/accounts/{quote(account_id, safe='')}/archive/export"
        response = self._client.post(path, request or {})
        return response.data or {}

    def entries(self, account_id: str, query: Optional[Dict[str, Any]] = None) -> EnumerationResult:
        """Enumerate archived entries for an account."""
        response = self._client.get(_append_query(f"/v1/archive/accounts/{quote(account_id, safe='')}/entries", query))
        if not response.data:
            return EnumerationResult()
        return EnumerationResult.from_dict(response.data, Entry.from_dict)

    def tenant_entries(
        self,
        tenant_id: str,
        account_id: str,
        query: Optional[Dict[str, Any]] = None
    ) -> EnumerationResult:
        """Enumerate archived entries for a tenant account."""
        path = f"/v1/tenants/{quote(tenant_id, safe='')}/accounts/{quote(account_id, safe='')}/entries"
        response = self._client.get(_append_query(path, query))
        if not response.data:
            return EnumerationResult()
        return EnumerationResult.from_dict(response.data, Entry.from_dict)

    def balance_as_of(self, account_id: str, as_of_utc: str) -> Dict[str, Any]:
        """Read archived balance as of a point in time for an account."""
        path = f"/v1/archive/accounts/{quote(account_id, safe='')}/balance/asof?asOf={quote(as_of_utc, safe='')}"
        response = self._client.get(path)
        return response.data or {}

    def tenant_balance_as_of(self, tenant_id: str, account_id: str, as_of_utc: str) -> Dict[str, Any]:
        """Read archived balance as of a point in time for a tenant account."""
        path = f"/v1/tenants/{quote(tenant_id, safe='')}/accounts/{quote(account_id, safe='')}/balance/asof?asOf={quote(as_of_utc, safe='')}"
        response = self._client.get(path)
        return response.data or {}

    def balance_asof(self, account_id: str, as_of_utc: str) -> Dict[str, Any]:
        """Read archived balance as of a point in time for an account."""
        return self.balance_as_of(account_id, as_of_utc)

    def tenant_balance_asof(self, tenant_id: str, account_id: str, as_of_utc: str) -> Dict[str, Any]:
        """Read archived balance as of a point in time for a tenant account."""
        return self.tenant_balance_as_of(tenant_id, account_id, as_of_utc)

    def verify_account(self, account_id: str, query: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Verify archived balance chain and object integrity for an account."""
        response = self._client.get(_append_query(f"/v1/archive/accounts/{quote(account_id, safe='')}/verify", query))
        return response.data or {}

    def verify_tenant_account(
        self,
        tenant_id: str,
        account_id: str,
        query: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Verify archived balance chain and object integrity for a tenant account."""
        path = f"/v1/tenants/{quote(tenant_id, safe='')}/accounts/{quote(account_id, safe='')}/verify"
        response = self._client.get(_append_query(path, query))
        return response.data or {}

    def object_metadata(self, object_id: str) -> Dict[str, Any]:
        """Read archive object catalog and storage metadata."""
        response = self._client.get(f"/v1/archive/objects/{quote(object_id, safe='')}/metadata")
        return response.data or {}

    def request_history(self, query: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Enumerate archived NetLedger request history from Archive Server."""
        response = self._client.get(_append_query("/v1/request-history", query))
        return response.data or {}

    def request_history_summary(self, query: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Summarize archived NetLedger request history from Archive Server."""
        response = self._client.get(_append_query("/v1/request-history/summary", query))
        return response.data or {}

    def request_history_entry(self, entry_id: str, query: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Read one archived NetLedger request history entry from Archive Server."""
        response = self._client.get(_append_query(f"/v1/request-history/{quote(entry_id, safe='')}", query))
        return response.data or {}

    def archive_server_request_history(self, query: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Enumerate Archive Server operational request history."""
        response = self._client.get(_append_query("/v1/archive-server/request-history", query))
        return response.data or {}

    def archive_server_request_history_summary(self, query: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Summarize Archive Server operational request history."""
        response = self._client.get(_append_query("/v1/archive-server/request-history/summary", query))
        return response.data or {}

    def archive_server_request_history_entry(self, entry_id: str, query: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Read one Archive Server operational request history entry."""
        response = self._client.get(_append_query(f"/v1/archive-server/request-history/{quote(entry_id, safe='')}", query))
        return response.data or {}
